compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion.hs

```diff
@@ -1011,6 +1007,124 @@ convertExpr env0 e = do
     go env o@(Coercion _) args = unhandled "Coercion" o
     go _ x args = unhandled "Expression" x
 
+-- | Is this an enum type?
+isEnumTyCon :: TyCon -> Bool
+isEnumTyCon tycon =
+    isEnumerationTyCon tycon
+    && (tyConArity tycon == 0)
+    && (length (tyConDataCons tycon) >= 2)
+
+-- | Is this a simple record type?
+isSimpleRecordTyCon :: TyCon -> Bool
+isSimpleRecordTyCon tycon =
+    maybe False isSimpleRecordCon (tyConSingleDataCon_maybe tycon)
+
+-- | Is this a variant type?
+isVariantTyCon :: TyCon -> Bool
+isVariantTyCon tycon =
+    (tyConFlavour tycon == DataTypeFlavour)
+    && not (isEnumTyCon tycon)
+    && not (isSimpleRecordTyCon tycon)
+
+conIsSingle :: DataCon -> Bool
+conIsSingle = isSingleConType . dataConTyCon
+
+conHasNoArgs :: DataCon -> Bool
+conHasNoArgs = null . dataConOrigArgTys
+
+conHasLabels :: DataCon -> Bool
+conHasLabels = notNull . ctorLabels
+
+isEnumCon :: DataCon -> Bool
+isEnumCon = isEnumTyCon . dataConTyCon
+
+isSimpleRecordCon :: DataCon -> Bool
+isSimpleRecordCon con = (conHasLabels con || conHasNoArgs con) && conIsSingle con
+
+isVariantRecordCon :: DataCon -> Bool
+isVariantRecordCon con = conHasLabels con && not (conIsSingle con)
+
+-- | The different classes of data cons with respect to LF conversion.
+data DataConClass
+    = EnumCon -- ^ constructor for an enum type
+    | SimpleRecordCon -- ^ constructor for a record type
+    | SimpleVariantCon -- ^ constructor for a variant type with no synthetic record type
+    | VariantRecordCon -- ^ constructor for a variant type with a synthetic record type
+    deriving (Eq, Show)
+
+classifyDataCon :: DataCon -> DataConClass
+classifyDataCon con
+    | isEnumCon con = EnumCon
+    | isSimpleRecordCon con = SimpleRecordCon
+    | isVariantRecordCon con = VariantRecordCon
+    | otherwise = SimpleVariantCon
+        -- in which case, daml-preprocessor ensures that the
+        -- constructor cannot have more than two unlabeled arguments
```

Comment:
The `SimpleVariantCon` case covers variant constructors whose fields are not labeled. Is that is correct, then I suggest
```suggestion
        -- constructor cannot have more than two arguments
```

Reply:
Yes, that looks good! Thanks!

Reply:
Changing this to `two or more arguments` which is more precise.